### PR TITLE
Enhancement/method annotation

### LINKED ISSUE #36: Add method level annotation in order to use circuit breaker
In order to ease the usage for the client an annotation could be provided that can be used on method level. The annotation will internally wrap the method call in the mentioned circuit breaker and use the configured fallback if necessary.

    @FailsafeBreaker(value="mybreaker", fallback="myFallback")
    public Reservation createOpenReservation(Reservation reservation) {
        ...
    }

### PR DESCRIPTION
**Pull request by coders-kitchen**
close #36

### CODE REVIEW

**Comment by MALPI** [comment]
Thanks for the PR @coders-kitchen, will have a look during the day!

File: src/main/java/org/zalando/failsafeactuator/service/FailsafeBreaker.java [hunk 1]
```diff
@@ -14,9 +14,7 @@
 import java.lang.annotation.Retention;
 import java.lang.annotation.Target;
 
-import static java.lang.annotation.ElementType.FIELD;
-import static java.lang.annotation.ElementType.METHOD;
-import static java.lang.annotation.ElementType.PARAMETER;
+import static java.lang.annotation.ElementType.*;
```
Comment (mastersimon) [inline src/main/java/org/zalando/failsafeactuator/service/FailsafeBreaker.java]: Please avoid using `*` imports.

File: src/main/java/org/zalando/failsafeactuator/aspect/FailsafeBreakerAspect.java [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+        try {
+          return executeFallback(methodInvocation, fallbackMethodName);
+        } catch (IllegalAccessException | IllegalArgumentException | InvocationTargetException fallbackException) {
+          throw fallbackException;
+        } catch (Exception fallbackException) {
+          log.debug("Fallback throws exception", fallbackException);
+        }
+      }
+
+      if (e.getCause() != null)
```
Comment (mastersimon) [inline src/main/java/org/zalando/failsafeactuator/aspect/FailsafeBreakerAspect.java]: Please add `{}`.

**Comment by MALPI** [comment]
Hey Peter LGTM just two things:

1.  Could you rename `FailsafeBreaker` annotation to `Failsafe`. This is shorter and from my point of view fitting better. What do you think?
2. I created the branch `release/0.6.0` can you file your PR once more against this branch? I'd like to add some more features to it before releasing the next version.

**Comment by MALPI** [comment]
Ah, just recognized there needs to be the possibility to reference the `retryPolicy` as well, right?

And I still thought about the option to make it more visible what's used as `Fallback` in the annotation. Maybe I am wrong, but wouldn't it be possible to use [functional interface](https://docs.oracle.com/javase/8/docs/api/java/util/function/Function.html)?

**Comment by coders-kitchen** [comment]
Hi Malte,

1. renaming sounds reasonable. Should we keep the existing annotation for compatibility reasons with the existing semantics? Or make it now simply ``@Failsafe``
2. Will do so

Sounds like a good idea to provide at least the _link_ to the retry policy. But maybe we could move this into a separate PR? WDYT?

In regards to the Fallback, the challenge here is that the fallback method has to have the same signature as the original method call (apart from the option that we could possible pass in the original exception as well). I'm not sure if this can be solved by a functional interface.

May be we could make it more explicit by annotating the methods with ``FailsafeFallback``which carries the circuit breaker name as parameter.

Using some of the lookup mechanics of spring we could possibly remove the restriction that the fallback method has to reside within the same class. By doing so, a client could extract an interface that is implemented twice. One has the original - primary - implementation  and the second could provide the fallback.

**Comment by coders-kitchen** [comment]
In regards to the fallback, I would propose to move this into a own PR as an improvement. WDYT?

File: src/main/java/org/zalando/failsafeactuator/service/FailsafeBreaker.java (hunk 1 above)
Comment (MALPI) [reply src/main/java/org/zalando/failsafeactuator/service/FailsafeBreaker.java]: Can you fix this one?

**Comment by MALPI** [comment]
We could keep both and mark it as `@Deprecated` until version 1.0. On the other hand I think it's not too much of an effort to switch to the other annotation and I'd like to avoid to carry around the old annotation for  a long time.

Since we are merging now against feature branch, I'm fine of moving fallback handling to another PR. 

The only thing to fix is the one `*` import.

**Comment by coders-kitchen** [comment]
* import is resolved

**Pull request merged**